tests/system/Database/Live/BadQueryTest.php

```diff
@@ -0,0 +1,28 @@
+<?php namespace CodeIgniter\Database\Live;
+
+use CodeIgniter\Test\CIDatabaseTestCase;
+
+/**
+ * @group DatabaseLive
+ */
+class BadQueryTest extends CIDatabaseTestCase
+{
+	protected $refresh = true;
+
+	protected $seed = 'Tests\Support\Database\Seeds\CITestSeeder';
+
+	public function testBadQuery()
+	{
+		if ($this->db->DBDebug) {
+			// expect an exception, class and message varies by DBMS
+			$this->expectException(\Exception::class);
+			$query = $this->db->query('SELECT * FROM table_does_not_exist');
+		} else {
+			// this throws an exception in this testing environment, but in production it'll return FALSE
+			// perhaps check $this->db->DBDebug for different test?
+			$query = $this->db->query('SELECT * FROM table_does_not_exist');
+			$this->assertEquals(false, $query);
+		}
```

Comment:
In test execution, the if condition is true or false, not both.
You should test both cases, `$this->db->DBDebug` is true and false.

Reply:
I still don't understand. If `$this->db-DBDEbug` casts as `true`, code execution starts at line 17 -- we must expect the exception before executing the query. Otherwise code execution goes to line 23, and we must run the query before we assert that its result is false.

Reply:
When we run the test case, we really test either of below:
- $this->expectException(\Exception::class);
- $this->assertEquals(false, $query);

But we should test both cases.
In other words, we should do:
1. set `$this->db->DBDebug` true and test
2. set `$this->db->DBDebug` false and test

Reply:
Aha! I think I understand now -- sorry.

I should remove the `if` statement and instead explicitly set `DBDebug=true` and test that and then set it to false and run that test also so that the test performs both tests every time. I was under the mistaken impression that the tests that ran would try it both ways.

Reply:
If I understand @kenjis and @michalsn correctly, you are saying I must construct these tests to *actually change the value of DBDebug* to force the test to run under both conditions. Unfortunately, **any change in this value appears to persist between tests and will affect other tests.** I.e., the unit tests start with whatever DBDebug default (default value, from Config, from .env, etc) and if you change DBDebug in one unit test method, that changed value will still be in effect for the following test.

Is there any function to restore the original DBDebug value? I suppose I might alter these unit tests to make note of the original value, but I need some additional information: Are all the tests in `DbUtilsTest` guaranteed to run in the order they are defined in the PHP file? Must I take care to restore the DBDebug value to its default before running each test?

NOTE: I tried defining a class property `$origDebug` and also defining a test to run first, `testFirst()` like so:
```php
	protected $origDebug;

	//--------------------------------------------------------------------

	/**
	 * This test must run first to store the inital debug value before we tinker with it below
	 */
	public function testFirst() {
		$this->origDebug = $this->getPrivateProperty($this->db, 'DBDebug');
		echo "orig in testFirst\n";
		var_dump($this->origDebug);
		
		$this->assertIsBool($this->origDebug);
	}
```
**HOWEVER** it would appear that phpunit instantiates a new instance of `DbUtilsTest` for each and every test method it contains so `$this->origDebug` will be `null` inside every other test function inside it even while any changes to `$this->db->DBDebug` persist between tests.

Please advise! Do I understand your instructions correctly? Is there a best practice for changing DBDebug values? Why are no other live db tests except DEbugTest manipulating this value?

Reply:
I have had some success if I make $origDebug a static variable because it then persists between different instances of `DBUtilsTest`:
```php
	protected static $origDebug;

	/**
	 * This test must run first to store the inital debug value before we tinker with it below
	 */
	public function testFirst() {
		$this::$origDebug = $this->getPrivateProperty($this->db, 'DBDebug');
		
		$this->assertIsBool($this::$origDebug);
	}
```
and then if I can be sure these tests will always run in the order they are defined in the PHP file, this should avoid changing DBDebug from whatever default it may have except for the scope of these tests:
```php
	public function testUtilsOptimizeTableFalseOptimizeDatabaseDebugTrue()
	{
		$util = (new Database())->loadUtils($this->db);
		$this->setPrivateProperty($util, 'optimizeTable', false);
		
		// set debug to true -- WARNING this change will persist!
		$this->setPrivateProperty($this->db, 'DBDebug', true);
		
		$this->expectException(DatabaseException::class);
		$this->expectExceptionMessage('Unsupported feature of the database platform you are using.');
		$util->optimizeDatabase();
		
		// this point in code execution will never be reached
	}

	public function testUtilsOptimizeTableFalseOptimizeDatabaseDebugFalse()
	{
		$util = (new Database())->loadUtils($this->db);
		$this->setPrivateProperty($util, 'optimizeTable', false);

		// set debug to false -- WARNING this change will persist!
		$this->setPrivateProperty($this->db, 'DBDebug', false);
		
		$result = $util->optimizeDatabase();
		$this->assertFalse($result);
		
		// restore original value grabbed from testFirst -- WARNING this change will persist!
		$this->setPrivateProperty($this->db, 'DBDebug', $this::$origDebug);
	}

```

Reply:
>  I.e., the unit tests start with whatever DBDebug default (default value, from Config, from .env, etc) and if you change DBDebug in one unit test method, that changed value will still be in effect for the following test.

Yes. Ideally, all the state before a test run should be completely identical. 

But currently, CI4 tests have a lot of hidden dependencies, and it will be difficult to control them perfectly,
so the best approach is to save the DBDebug value before execution and restore it.

Reply:
> HOWEVER it would appear that phpunit instantiates a new instance of DbUtilsTest for each and every test method it contains so $this->origDebug will be null inside every other test function inside it even while any changes to $this->db->DBDebug persist between tests.

PHPUnit resets properties in a test case, because the state before each test should be identical
and ideally a test should not depend on another test.

If you want to keep a value between test methods, you can use static properties or Test Dependencies
https://phpunit.readthedocs.io/en/9.5/writing-tests-for-phpunit.html#test-dependencies

Reply:
> PHPUnit resets properties in a test case, because the state before each test should be identical
> and ideally a test should not depend on another test.

I did test whether a change to the DBDebug value in one test method would persist when PHPUnit moves on to the next test and it did ⚠️⚠️⚠️.

I tried adding a class property that was non-static, `public $origDebug` and I set this value in a test method but, to my surprise, it was NULL in every subsequent test method in that file. This is what made me think it was creating a new instance of the test class for each method. I added a constructor function and this was called repeatedly for a single test file -- presumably once for each test. I surmised from these experiments that the db object would retain any changes made to DBDebug until *all* the subsequent tests had run. PHP passes objects by reference, so if you modify them inside a function, the changes can persist:
```php
class foo {
  public $prop = "default";
}

function myFunc($v) {
  $v->prop = "inside";
  echo $v->prop . "\n";
}

$v = new foo;
echo $v->prop . "\n";

myFunc($v);

echo $v->prop . "\n";
```
output:
```
default
inside
inside
```
 
> If you want to keep a value between test methods, you can use static properties or Test Dependencies
> https://phpunit.readthedocs.io/en/9.5/writing-tests-for-phpunit.html#test-dependencies

I did indeed end up using a `public static $origDebug` var to hold the original/default value for DBDebug, and this seems to work, but I would point out that *this assumes the tests will be executed in the order they are defined in the file*.

Reply:
> I did test whether a change to the DBDebug value in one test method would persist when PHPUnit moves on to the next test and it did ⚠️⚠️⚠️.

`$this->db` comes from the static method (it was moved to trait recently):

CodeIgniter4/DatabaseTestTrait.php at develop · codeigniter4/CodeIgniter4
https://github.com/codeigniter4/CodeIgniter4/blob/e4661270a2b4ce12bbc2ec456ea984e3f149ec36/system/Test/DatabaseTestTrait.php#L72-L76

> this assumes the tests will be executed in the order they are defined in the file.

If I remember correctly, phpunit runs tests in the order.